src/analyser.cpp

```diff
@@ -0,0 +1,1523 @@
+                                           })
+                                           != std::end(mInternalVariables);
+
+        if (hasUnderconstrainedVariables) {
+            if (hasOverconstrainedVariables) {
+                mModel->mPimpl->mType = AnalyserModel::Type::UNSUITABLY_CONSTRAINED;
+            } else {
+                mModel->mPimpl->mType = AnalyserModel::Type::UNDERCONSTRAINED;
+            }
+        } else if (hasOverconstrainedVariables) {
+            mModel->mPimpl->mType = AnalyserModel::Type::OVERCONSTRAINED;
+        } else if (mModel->mPimpl->mVoi != nullptr) {
+            mModel->mPimpl->mType = AnalyserModel::Type::ODE;
+        } else if (!mInternalVariables.empty()) {
+            mModel->mPimpl->mType = AnalyserModel::Type::ALGEBRAIC;
+        }
+    } else {
+        mModel->mPimpl->mType = AnalyserModel::Type::INVALID;
+    }
+
+    // Some final post-analysis is now needed, if we have a valid model.
+
+    if ((mModel->mPimpl->mType == AnalyserModel::Type::ODE)
+        || (mModel->mPimpl->mType == AnalyserModel::Type::ALGEBRAIC)) {
+        // Sort our internal variables and equations.
+
+        std::sort(mInternalVariables.begin(), mInternalVariables.end(),
+                  compareVariablesByTypeAndIndex);
+        std::sort(mInternalEquations.begin(), mInternalEquations.end(),
+                  compareEquationsByVariable);
+
+        std::map<VariablePtr, AnalyserEquationPtr> equationMappings;
+        std::map<AnalyserEquationPtr, AnalyserVariablePtr> variableMappings;
+
+        for (const auto &internalEquation : mInternalEquations) {
+            equationMappings[internalEquation->mVariable->mVariable] = std::shared_ptr<AnalyserEquation> {new AnalyserEquation {}};
+        }
+
+        // Make our internal variables available through our API.
+
+        auto stateIndex = MAX_SIZE_T;
+        auto variableIndex = MAX_SIZE_T;
+
+        for (const auto &internalVariable : mInternalVariables) {
+            // Determine the type of the variable.
+
+            AnalyserVariable::Type type;
+
+            if (internalVariable->mType == AnalyserInternalVariable::Type::STATE) {
+                type = AnalyserVariable::Type::STATE;
+            } else if (internalVariable->mType == AnalyserInternalVariable::Type::CONSTANT) {
+                type = AnalyserVariable::Type::CONSTANT;
+            } else if ((internalVariable->mType == AnalyserInternalVariable::Type::COMPUTED_TRUE_CONSTANT)
+                       || (internalVariable->mType == AnalyserInternalVariable::Type::COMPUTED_VARIABLE_BASED_CONSTANT)) {
+                type = AnalyserVariable::Type::COMPUTED_CONSTANT;
+            } else if (internalVariable->mType == AnalyserInternalVariable::Type::ALGEBRAIC) {
+                type = AnalyserVariable::Type::ALGEBRAIC;
+            } else {
+                // This is the variable of integration, so skip it.
+
+                continue;
+            }
+
```

Comment:
??

Reply:
The final continue statement.  Why is it needed? Why not just remove the `else`?

Reply:
Are you suggesting deleting lines 1381 to 1384? If so, then we definitely don't want to do that. We check the type of an internal variable and if it is the VOI then we really don't want to do what follows (i.e. populate and keep track of the variable in `mModel`). This is why we call `continue` so that it goes back to the beginning of the `for` loop.